analytics/transform_test.go

```diff
@@ -202,13 +202,11 @@ func TestToJson(t *testing.T) {
 	assert := assert.New(t)
 
 	// correct value
-	jsonEvent, err := stdJson.Marshal(eventMapWithoutGeo)
-	if err != nil {
-	}
+	jsonEvent, _ := jsoniter.Marshal(eventMapWithoutGeo)
 
 	jsonifiedEvent, err := fullEvent.ToJson()
 	assert.Nil(err)
-	assert.Equal(jsonEvent, jsonifiedEvent)
+	assert.Equal(len(jsonEvent), len(jsonifiedEvent))
```

Comment:
Since now we are using the default jsoniter configuration, we are not sorting the map keys anymore. This means that asserting the equality of fullEvent.ToJson() and jsoniter.Marshal() is not possible anymore since the keys will be random. One workaround would be to get both []byte outputs and convert them to maps but that overly complicates the unit tests. The other workaround, which is the one I implemented, is testing that both []byte outputs have the same length. This does not guarantee that they will be equal, but it's better than nothing.

Reply:
The old way tested for exact equality between the two `bytes[]` objects, I don't think comparing lengths can be an improvement on that. `test` and `pest` have the same lengths after being converted to bytes (`74 65 73 74` vs `70 65 73 74`) - the new test would say they're the same.

I think it'll be helpful if I try to explain the problem - I don't think we should be particularly concerned with asserting that `fullevent.ToJson()` and `jsoniter.Marshal()` (or `json.Marshal()` for that matter) have the same result, especially since one uses the other.

jsoniter's default behaviour is the same as `json`'s in this respect, so actually using one or the other library makes no difference. The problem which led to a bug (and made it difficult to spot the bug) is that we're transforming a map to JSON on both sides of the assertion, and both of them had the same (unexpected) relevant behaviour (not escaping unicode where I assumed it had been escaped).

We can remove the possiblity of that problem, and make the tests more explicit, by hardcoding exactly the output we'd like to see as a bytes-encoded JSON string. That way, our test case must look like this:

```
[]byte{"{... "app_id":"\u003c\u003eangry-birds" ...}"}
```

Instead, we had it defined in a map which used `<>angry-birds`, and then we converted that map to JSON. I made the mistake of thinking that I was asserting against `<>` but after it was marshalled, I was actually asserting against `\u003c\u003e` - the point is that the way I originally wrote the tests makes it too easy to make such a mistake. 

I hope that makes sense?

Reply:
Ok, so I found a way to test the equality of the []bytes. While the default jsoniter configuration that we use does not sort the JSONs it Marshals, the standard json library does, so I just added a simple Remarshal function that returns the []byte JSON sorted and use it for comparing the outputs in tests.

```go
jsonEventOrdered, _ := orderJson(jsonEvent)
	jsonifiedEventOrdered, _ := orderJson(jsonifiedEvent)

	assert.Equal(jsonEventOrdered, jsonifiedEventOrdered)
```

Reply:
So my issue with the test is that we shouldn't be comparing to a marshalled map, we should be comparing to a hardcoded expected value - like we do in stream replicator: https://github.com/snowplow-devops/stream-replicator/blob/38160a0f3f0a5cfd7e9e3daa8e9d7536ff50e145/pkg/transform/transform_test_variables.go#L16

The reasoning is that when we expected escaped unicode, the value we were comparing against appeared to contain `<>`, but after marshalling, it was actually ` \u003c\u003e` - and so it wasn't clear that we weren't escaping unicode when we thought we were.

Reply:
I finally got it, sorry that took a bit. I've created constants with unescaped unicode and I'm comparing to them now.